clang/lib/CIR/Dialect/IR/CIRAttrs.cpp

```diff
@@ -295,6 +295,104 @@ LogicalResult IntAttr::verify(function_ref<InFlightDiagnostic()> emitError,
   return success();
 }
 
+//===----------------------------------------------------------------------===//
+// DataMemberAttr definitions
+//===----------------------------------------------------------------------===//
+
+Attribute DataMemberAttr::parse(AsmParser &parser, Type type) {
+  auto loc = parser.getCurrentLocation();
+
+  if (parser.parseLess())
+    return {};
+
+  Type clsTy;
+  std::int64_t memberIndex;
+
+  if (parser.parseOptionalKeyword("null").succeeded()) {
+    clsTy = type.cast<mlir::cir::DataMemberType>().getClsTy();
+    memberIndex = -1;
+  } else {
+    if (parser.parseType(clsTy))
+      return {};
+
+    if (parser.parseComma())
+      return {};
+
+    if (parser.parseInteger(memberIndex))
+      return {};
+  }
+
+  if (parser.parseGreater())
+    return {};
+
+  auto memberIndexTy = IndexType::get(parser.getContext());
+  auto memberIndexAttr = IntegerAttr::get(memberIndexTy, memberIndex);
+
+  return getChecked([&parser, loc] { return parser.emitError(loc); },
+                    parser.getContext(), type, clsTy, memberIndexAttr);
+}
+
+void DataMemberAttr::print(AsmPrinter &printer) const {
+  printer << '<';
```

Comment:
Both parser and printer seem simple enough that they could be written directly in tablegen using declarative syntax, did you run into any problems while trying that approach?

Reply:
Updated.